### PR TITLE
Add some base structure to reduce test set-up time

### PR DESCRIPTION
**Pull request by glennandskinandme**
Added basic flask app, routes, models, etc to reduce the set-up time for candidates doing the test.

### CODE REVIEW

File: app/application.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    order = None
+    try:
+        payload = request.json
+    except Exception as e:
+        return _standard_response(str(e), 500)
+
+    with app.app_context():
+
+        customer = Customer.query.filter_by(id=payload['customer']).one_or_none()
```
Comment (tfranceuk) [inline app/application.py]: use of `query` directly on `Customer` is not a clean pattern.
I would use a session to query against.
Probably in this case `db.session.query`

File: app/application.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@app.route('/')
+def index():
+    # Default route - blank page
+    return render_template('index.html')
+
+
+@app.route('/order', methods=['POST'])
+def order_product():
```
Comment (tfranceuk) [inline app/application.py:33]: Should this file be purely about exposing app instance to be ran?
And so should routes like this live in `application/routes/orders` or something?

Or are we looking for feedback on this sort of thing too as part of this test?

**Review by tfranceuk — COMMENTED** [review]
Hey Glenn,

Not sure how much critiquing I should do here, so only added a couple of inline comments on code.

I also have the following comments generally:

- Readme
	* Typo: “The fictinional billing provider” > “The fictional billing provider”
	* Db diagram has 1:M in wrong direction for Customer:Order
	* No detail on prerequisites, e.g. Python ^3.8
	* No detail on setting up environment, e.g. setup venv, pip install requirements, etc.
	* .gitignore .venv, if this is suggested env
	* how to run tests
		* had to `cd app`, then `python -m pytest tests`
	* how to create database
	* how to run flask app
- Might be good to ask for feedback on our code, and can discuss during interview.  Always good to discuss different styles/approaches.
- Tests don’t run: `MySQLdb._exceptions.OperationalError: (2005, "Unknown MySQL server host 'database_server' (0)”)`
	* Used connection string: `mysql://root:my-secret-pw@127.0.0.1:33060/assessment?charset=utf8mb4`
	* `cd app` > `python create-database.py`
		* Error when creating customer from customer_factory: `'NoneType' object has no attribute 'encode'`

File: app/application.py (hunk 1 above)
Comment (glennandskinandme) [reply app/application.py]: Correct and fixed

File: app/application.py (hunk 2 above)
Comment (glennandskinandme) [reply app/application.py:33]: I have offered the candidate the opportunity to 'tidy' things up a bit. This is an obvious one to fix.

File: README.md [hunk 3]
```diff
@@ -50,17 +50,25 @@ Please mock up (in Python/Flask/SQLAlchemy) the schema with columns that you fee
...
 * alter the database to support the new functionality adding columns and tables as required.
+* restructure any elements of the app to make it a little more standard/well structured.
 
 ## Scope
 **This is obviously a broad task. The intention is to allow you to demonstrate a variety of skills to the degree you feel comfortable. Any elements of the system you don't implement can be annotated so we can get an understanding of what you would have done, given time and opportunity**. This is not intended to be a robust production system. The purpose of the assessment is to show your Python, Flask and SQLAlchemy skills. Please also see [Hints and Tips](/#hints-and-tips), below.
 
+# Prerequisites
+- Python 3.8
+- Flask v1
+- Virtualenv
```
Comment (tfranceuk) [inline README.md]: These aren't prerequisites in my opinion.
They will be installed as part of the setup.

Comment (tfranceuk) [reply README.md]: On that note, can we give a setup section:

### Setup
* `python -m venv .venv`
* `source .venv/bin/activate`
* `pip install -r app/requirements.txt`
#### Running server
* `(cd app && FLASK_APP=application:app FLASK_ENV=development flask run --host 0.0.0.0 --port 9000)`
#### Running tests
* `(cd app && python -m pytest tests)`

Comment (tfranceuk) [reply README.md]: and finally, maybe mention changing the value of `app.config['SQLALCHEMY_DATABASE_URI']` in `app/assessment/app.py` to their mysql db (or `sqlite:///file.db`)

**Review by tfranceuk — APPROVED** [review]
Hey Glenn,

Have added some suggestions for additions to project setup in readme.
See what you think.

Thanks

File: README.md (hunk 3 above)
Comment (glennandskinandme) [reply README.md]: All good comments/observations - have updated the README (incorporating the app/README.md that wasn't added to the PR) and added the item related to SQLite which is a good observation.

**Pull request merged**